### PR TITLE
Chore - Implements Cache on V3

### PR DESCRIPTION
**Pull request by djonasm**

### CODE REVIEW

File: src/Model/ModelInterface.php [hunk 1]
```diff
@@ -22,6 +22,7 @@ interface ModelInterface extends HasAttributesInterface, Persistable
      *
      * @param array $query      mongoDB selection criteria
      * @param array $projection fields to project in Mongo query
+     * @param bool  $useCache   retrieves a CacheableCursor instead
      */
     public static function where(array $query = [], array $projection = []): CursorInterface;
```
Comment (ezandonai) [inline src/Model/ModelInterface.php]: Faltou o parametro aqui jovem.
```suggestion
    public static function where(array $query = [], array $projection = [], bool $useCache = false): CursorInterface;
```

File: tests/Unit/Cursor/CacheableCursorTest.php [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+class CacheableCursorTest extends TestCase
+{
+    public function tearDown(): void
+    {
+        parent::tearDown();
+        m::close();
+    }
+
+    public function testShouldGetCursorFromPreviousIteration()
```
Comment (ezandonai) [inline tests/Unit/Cursor/CacheableCursorTest.php]:
```suggestion
    public function testShouldGetCursorFromPreviousIteration(): void
```

File: tests/Unit/Cursor/CacheableCursorTest.php [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        );
+
+        // Assert
+        $this->assertEquals(
+            new ArrayIterator($documentsFromDb),
+            $this->callProtected($cursor, 'getCursor')
+        );
+    }
+
+    public function testShouldGetCursorFromCache()
```
Comment (ezandonai) [inline tests/Unit/Cursor/CacheableCursorTest.php]:
```suggestion
    public function testShouldGetCursorFromCache(): void
```

File: tests/Unit/Cursor/CacheableCursorTest.php [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+            ->andReturn($documentsFromCache);
+
+        // Assert
+        $this->assertEquals(
+            new ArrayIterator($documentsFromCache),
+            $this->callProtected($cursor, 'getCursor')
+        );
+    }
+
+    public function testShouldGetFromDatabaseWhenCacheFails()
```
Comment (ezandonai) [inline tests/Unit/Cursor/CacheableCursorTest.php]:
```suggestion
    public function testShouldGetFromDatabaseWhenCacheFails(): void
```

File: tests/Unit/Cursor/CacheableCursorTest.php [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+            ->with($cacheKey, $documentsFromDb, m::any());
+
+        // Assert
+        $this->assertEquals(
+            new ArrayIterator($documentsFromDb),
+            $this->callProtected($cursor, 'getCursor')
+        );
+    }
+
+    public function testShouldGetCursorFromDatabaseAndCacheForLater()
```
Comment (ezandonai) [inline tests/Unit/Cursor/CacheableCursorTest.php]:
```suggestion
    public function testShouldGetCursorFromDatabaseAndCacheForLater(): void
```

File: tests/Unit/Cursor/CacheableCursorTest.php [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+            ->with('find:collection:123', $documentsFromDb, m::any());
+
+        // Assert
+        $this->assertEquals(
+            new ArrayIterator($documentsFromDb),
+            $this->callProtected($cursor, 'getCursor')
+        );
+    }
+
+    public function testShouldGetOriginalCursorFromDatabaseAfterTheDocumentLimit()
```
Comment (ezandonai) [inline tests/Unit/Cursor/CacheableCursorTest.php]:
```suggestion
    public function testShouldGetOriginalCursorFromDatabaseAfterTheDocumentLimit(): void
```

File: tests/Unit/Cursor/CacheableCursorTest.php [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+            ->never();
+
+        // Assert
+        $this->assertEquals(
+            (new ArrayIterator($documentsFromDb))->current(),
+            $this->callProtected($cursor, 'getCursor')->current()
+        );
+    }
+
+    public function testShouldGenerateUniqueCacheKey()
```
Comment (ezandonai) [inline tests/Unit/Cursor/CacheableCursorTest.php]:
```suggestion
    public function testShouldGenerateUniqueCacheKey(): void
```

File: tests/Unit/Cursor/CacheableCursorTest.php [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+
+        $expectedCacheKey = sprintf(
+            '%s:%s:%s',
+            'find',
+            'my_db.my_collection',
+            md5(serialize([['color' => 'red']]))
+        );
+
+        // Assert
+
```
Comment (ezandonai) [inline tests/Unit/Cursor/CacheableCursorTest.php]:
```suggestion
```

File: tests/Unit/Util/CacheComponentTest.php [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+     */
+    public $time = 1466710000;
+
+    public function tearDown(): void
+    {
+        parent::tearDown();
+        m::close();
+    }
+
+    public function testShouldImplementCacheComponentInterface()
```
Comment (ezandonai) [inline tests/Unit/Util/CacheComponentTest.php]:
```suggestion
    public function testShouldImplementCacheComponentInterface(): void
```

File: tests/Unit/Util/CacheComponentTest.php [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    {
+        // Arrange
+        $cacheComponent = (new CacheComponent());
+
+        // Assertion
+        $this->assertInstanceOf(CacheComponentInterface::class, $cacheComponent);
+    }
+
+    public function testShouldPutAndRetrieveValues()
```
Comment (ezandonai) [inline tests/Unit/Util/CacheComponentTest.php]:
```suggestion
    public function testShouldPutAndRetrieveValues(): void
```

File: tests/Unit/Util/CacheComponentTest.php [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        $cacheComponent = $this->getCacheComponent();
+
+        // Assertion
+        $cacheComponent->put('saveThe', 'bacon', 1); // 1 minute of ttl
+        $this->tick(30); // After 30 seconds
+        $this->assertTrue($cacheComponent->has('saveThe'));
+        $this->assertEquals('bacon', $cacheComponent->get('saveThe'));
+    }
+
+    public function testShouldExpireValues()
```
Comment (ezandonai) [inline tests/Unit/Util/CacheComponentTest.php]:
```suggestion
    public function testShouldExpireValues(): void
```

**Review by ezandonai — APPROVED** [review]

**Review by orlandocavassani — APPROVED** [review]

File: src/Util/CacheComponentInterface.php [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    public function get(string $key);
+
+    /**
+     * Store an item in the cache for a given number of minutes.
+     *
+     * @param string $key     cache key of the item
+     * @param mixed  $value   value being stored in cache
+     * @param float  $minutes cache ttl
+     */
+    public function put(string $key, $value, float $minutes);
```
Comment (orlandocavassani) [inline src/Util/CacheComponentInterface.php:27]:
```suggestion
    public function put(string $key, $value, float $seconds);
```
There are differences between `mongolid` and `mongolid-laravel`. In `mongolid` the cache is handled in minutes and `mongolid-laravel` is handled in seconds.

I think we could use the same behavior on both.

Comment (ezandonai) [reply src/Util/CacheComponentInterface.php:27]: This part of cache is actually a rollback. Cause in the v3 doesnt have cache.
If in the future we decide to keep with cache, will considerate to make this change.

**Pull request merged**